core/src/jmh/java/org/apache/iceberg/ReplaceDeleteFilesBenchmark.java

```diff
@@ -104,27 +108,39 @@ private void dropTable() {
     TABLES.dropTable(TABLE_IDENT);
   }
 
-  private void initFiles() {
+  private void initFiles() throws IOException {
     List<DeleteFile> generatedDeleteFiles = Lists.newArrayListWithExpectedSize(numFiles);
     List<DeleteFile> generatedPendingDeleteFiles = Lists.newArrayListWithExpectedSize(numFiles);
 
     RowDelta rowDelta = table.newRowDelta();
+    int filesToDelete = (int) Math.ceil(numFiles * (percentDeleteFilesReplaced / 100.0));
 
     for (int ordinal = 0; ordinal < numFiles; ordinal++) {
       DataFile dataFile = FileGenerationUtil.generateDataFile(table, null);
       rowDelta.addRows(dataFile);
-
-      DeleteFile deleteFile = FileGenerationUtil.generatePositionDeleteFile(table, dataFile);
-      rowDelta.addDeletes(deleteFile);
-      generatedDeleteFiles.add(deleteFile);
-
-      DeleteFile pendingDeleteFile = FileGenerationUtil.generatePositionDeleteFile(table, dataFile);
-      generatedPendingDeleteFiles.add(pendingDeleteFile);
+      if (filesToDelete > 0) {
+        DeleteFile deleteFile = FileGenerationUtil.generatePositionDeleteFile(table, dataFile);
+        rowDelta.addDeletes(deleteFile);
+        generatedDeleteFiles.add(deleteFile);
+        DeleteFile pendingDeleteFile =
+            FileGenerationUtil.generatePositionDeleteFile(table, dataFile);
+        generatedPendingDeleteFiles.add(pendingDeleteFile);
+        filesToDelete--;
```

Comment:
Opting for keeping the benchmarking pretty simple for now. We just use the percentage and this will correlate linearly with the amount of manifests which need to get rewritten. We could test cases where we vary the distribution of deletes across partitions, but don't think it's strictly required now since the current benchmark does indicate this is a good improvement.

Reply:
If I understand correctly, the new and the old logic replace all delete files. Your PR would benefit use cases when a small number of manifests has to be rewritten, so I'd probably add `numFiles` data and delete files but replace only `percentDeleteFilesReplaced`.

Also, I would consider running the same benchmark for an unpartitioned table as it will not have the partition filters and Iceberg would be forced to scan through all of the metadata even if a single delete file is replaced.

Reply:
I updated the benchmark to only replace a percentage of the delete files rather than before where the percentage was more a ratio of how many data files had delete files.  That's probably a more useful bench for evaluating synchronous maintenance since we get a sense of how expensive the operation is at varying levels of replacement.

After updating to have the set be a source of truth for delete manifests, a more noticeable difference is observed, as expected! I updated the benchmark results in the PR description.